### PR TITLE
[inductor] Add lowering for aten.take

### PR DESCRIPTION
**Pull request by brenocfg**
Adds lowering for `aten.take`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel @lezcano

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -1019,6 +1019,32 @@ def diagonal_scatter(input, src, offset: int = 0, dim1: int = 0, dim2: int = 1):
...
+    flat = view(x, [sympy_product(x.get_size())])
+    size = flat.get_size()
+    flat_loader = flat.make_loader()
+    index_loader = index.make_loader()
+
+    def fn(idx):
+        idx = list(idx)
+        idx = index_loader(idx)
+        # deal with negative indices
+        idx = ops.where(ops.lt(idx, 0), idx + ops.constant(size[0], torch.int64), idx)
```
Comment (lezcano) [inline torch/_inductor/lowering.py:1037]: Negative indices are dealt with later on, no need to add this here.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -1019,6 +1019,32 @@ def diagonal_scatter(input, src, offset: int = 0, dim1: int = 0, dim2: int = 1):
...
+    assert isinstance(index, TensorBox)
+    assert "int" in str(index.get_dtype())
+
+    flat = view(x, [sympy_product(x.get_size())])
+    size = flat.get_size()
+    flat_loader = flat.make_loader()
+    index_loader = index.make_loader()
+
+    def fn(idx):
+        idx = list(idx)
```
Comment (lezcano) [inline torch/_inductor/lowering.py:1034]: Do we need this cast?

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (lezcano) [reply torch/_inductor/lowering.py:1037]: Ah, right, negative indices should fail on CPU, but not on GPU. We can xfail those for now. Those will work once https://github.com/pytorch/pytorch/pull/108690 is merged.

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (brenocfg) [reply torch/_inductor/lowering.py:1034]: Oops, we do not!

**Review by lezcano — APPROVED** [review]
Great, thank you! In retrospect, this could have been a decomposition in terms of flatten + advanced indexing. Perhaps you could implement `put` that way as it may be easier?

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -1019,6 +1019,29 @@ def diagonal_scatter(input, src, offset: int = 0, dim1: int = 0, dim2: int = 1):
...
+    assert "int" in str(index.get_dtype())
+
+    flat = view(x, [sympy_product(x.get_size())])
+    size = flat.get_size()
+    flat_loader = flat.make_loader()
+    index_loader = index.make_loader()
+
+    def fn(idx):
+        idx = [ops.indirect_indexing(index_loader(idx), size[0])]
+        return flat_loader(idx)
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:1035]: Could this be a decomposition?
```
@register_decomposition(aten.take)
def take(x, index):
    return x.reshape(-1).gather(dim=0, index=index)
```

Comment (lezcano) [reply torch/_inductor/lowering.py:1035]: Yep, that + out_wrapper. See also https://github.com/pytorch/pytorch/pull/109784#pullrequestreview-1638289609

This was an onboarding task to get familiarized with the PyTorch IR, but it's true that it'd be better to write it as a decomp. In the decomp we could also add some checks that may be missing here.

**Comment by lezcano** [comment]
fwiw, on a second thought, I think that `put` may not be possible to implement as a decomposition really.

**Pull request closed without merging**